File: python/2015/day05b/day05b.py
#!/usr/bin/env python3

import re
import sys

def usage():
    progname = sys.argv[0]
    print(f'usage: {progname} <input file>')
    sys.exit(1)

def prop1(line):
    return bool(re.search(r'(..).*\1', line))

def prop2(line):
    return bool(re.search(r'(.).\1', line))

def process(filename):
    counter = 0
    with open(filename, 'r') as infile:
        for line in infile:
            if prop1(line) and prop2(line):
                counter += 1
    return counter

def main():
    if len(sys.argv) < 2:
        usage()
    filename = sys.argv[1]
    result = process(filename)
    print(f'result = {result}')

if __name__ == '__main__':
    main()

File: perl/2015/day05b/day05b.pl
#!/usr/bin/env perl

use strict;
use warnings;

sub usage {
    my $progname = $0;
    print "usage: $progname <input file>\n";
    exit 1;
}

sub prop1 {
    my ($str) = @_;
    return ($str =~ /(..).*\1/);
}

sub prop2 {
    my ($str) = @_;
    return ($str =~ /(.).\1/);
}

sub process {
    my ($filename) = @_;
    my $counter = 0;

    open ( my $infile, '<', $filename ) or die "cannot open file: $filename: $!";
    while ( my $line = <$infile> ) {
        if ( prop1($line) && prop2($line) ) {
            $counter++;
        }
    }
    close($infile);

    return $counter;
}

sub main {
    if ( @ARGV < 1 ) {
        usage();
    }
    my $filename = $ARGV[0];
    my $result = process($filename);
    print "result = $result\n";
}

main();
